### PR TITLE
Reorganized coding effects to use KnownAminoAcidChange base class

### PR DESCRIPTION
**Pull request by iskandr**
Also added unit test for NoncodingTranscript effect

---
This change is [Reviewable](https://reviewable.io/reviews/hammerlab/varcode/153)

### CODE REVIEW

File: test/test_effect_classes.py [hunk 1]
```diff
@@ -75,7 +88,9 @@ def test_start_loss():
     expect_effect(
         variant,
         transcript_id="ENST00000420316",
-        effect_class=StartLoss)
+        effect_class=StartLoss,
+        modifies_coding_sequence=True,
+        modifies_protein_sequence=True)
```
Comment (arahuja) [inline test/test_effect_classes.py]: `StartLoss` modifies_protein_sequence? I guess it does, but tbh not what I would have expected when filtering on `modifies_protein_sequence` so good to know!

File: varcode/effects.py [hunk 2]
```diff
@@ -382,13 +395,42 @@ def __init__(
         self.aa_mutation_end_offset = aa_mutation_end_offset
         self.aa_ref = aa_ref
 
-    modifies_protein_sequence = True
+    @property
+    def modifies_protein_sequence(self):
```
Comment (arahuja) [inline varcode/effects.py]: In the tests you have 

```       
         effect_class=AlternateStartCodon,
         modifies_coding_sequence=True,
         modifies_protein_sequence=False)
```

File: varcode/effects.py [hunk 3]
```diff
@@ -382,13 +395,42 @@ def __init__(
...
+        # TODO: scan downstream in the cDNA sequence to predict an alternate
+        # start codon using a Kozak sequence template
+        return None
+
+    @memoized_property
+    def short_description(self):
+        return "p.%s1? (start-loss)" % (self.transcript.protein_sequence[0],)
 
-class BaseSubstitution(NonsilentCodingMutation):
+class KnownAminoAcidChange(NonsilentCodingMutation):
```
Comment (arahuja) [inline varcode/effects.py]: Does the `Known` part imply anything here? Why not just `AminoAcidChange`? This is everything that has `modifies_protein_sequence == True`?

Comment (iskandr) [inline varcode/effects.py]: Effects such as `StartLoss` modify the protein sequence but we don't know what the amino acid change will be! So, this is really a base class for any effect with non-`None` mutant protein sequence.

File: varcode/effects.py (hunk 2 above)
Comment (iskandr) [inline varcode/effects.py]: Good catch, strange that the unit test doesn't fail.

Comment (iskandr) [inline varcode/effects.py]: Actually, I don't see that on my local copy of the file.

File: test/test_effect_classes.py (hunk 1 above)
Comment (iskandr) [inline test/test_effect_classes.py]: I'm imagining any variant you would expect to give you a different protein (even if its sequence is unknown) has `modifies_protein_sequence` as True. The only weird case I haven't figured out yet is `ExonLoss`.

File: varcode/effects.py (hunk 2 above)
Comment (iskandr) [inline varcode/effects.py]: Oh, the diff is confusingly mixing up the head of `AlternateStartCodon` and the tail of `NonsilentCodingMutation`

**Pull request merged**